File: day9/puzzle.go
// Package day9 has solutions for Day 9 of Advent of Code
// https://adventofcode.com/2020/day/9
package day9

import (
	"fmt"

	"github.com/rthorpeii/AdventOfCode2020/input"
)

var inputFile string = "day9/input.txt"
var testFile string = "day9/testInput.txt"

// SolvePuzzle prints the output produced by running the input and test files on both parts
func SolvePuzzle() {
	fmt.Printf("Part 1 - Test: %v \n", PartOne(testFile))
	fmt.Printf("Part 1 - Actual: %v \n", PartOne(inputFile))
	fmt.Printf("Part 2 - Test: %v \n", PartTwo(testFile))
	fmt.Printf("Part 2 - Actual: %v \n", PartTwo(inputFile))
}

// What size should the preamble f
var preamble = 25

// PartOne finds what number in the input is not the sum of two of the previous 25 numbers
func PartOne(file string) int {
	input := input.IntSlice(file)
	return findTarget(input, 25)
}

// Finds what number in the input is not the sum of the previous x numbers
// where x is preamble
func findTarget(input []int, preamble int) int {
	for i := preamble; i < len(input); i++ {
		if !valid(input[i-preamble:i], input[i]) {
			return input[i]
		}
	}
	return -1
}

// valid determines if the target can be created by adding two numbers within the input
func valid(input []int, target int) bool {
	for first, val := range input {
		for second := first + 1; second < len(input); second++ {
			if val+input[second] == target {
				return true
			}
		}
	}

	return false
}

// PartTwo finds what consecutive sequence of at least 2 numbers in the input sum to the
// target number which is the number in the input that is not the sum of two of the previous 25 numbers
func PartTwo(file string) int {
	rawInput := input.IntSlice(file)
	preamble := 25
	target := findTarget(rawInput, preamble)

	min, max := 0, 1
	sum := rawInput[min]
	for max < len(rawInput)-1 {
		sum += rawInput[max]

		for sum > target {
			if max-min == 0 {
				break
			}
			sum -= rawInput[min]
			min++
		}

		if sum == target {
			break
		}

		max++
	}

	smallest, largest := rawInput[max], rawInput[max]
	for index := min; index < max; index++ {
		val := rawInput[index]
		if val < smallest {
			smallest = val
		} else if val > largest {
			largest = val
		}
	}

	return smallest + largest
}

File: 2021/day01/puzzle.rb
# Solutions for Day 1 of Advent of Code
# https://adventofcode.com/2021/day/1
require "../input.rb"

puzzleInputFile = "input.txt"
testInputFile = "testInput.txt"

# Counts the number of input lines that are larger than the previous input line
def partOne (inputFile) 
    input = parseAsIntArray(inputFile)
    
    counter = 0
    input.each_cons(2) do |a,b|
        if b > a
            counter += 1
        end 
    end
    return counter
end

# Counts the number of consecutive chunks of 3 input lines that have a higher sum than that
# of the previous chunk of 3 consecutive input lines.
def partTwo(inputFile)
    input = parseAsIntArray(inputFile)
    
    counter = 0
    input.each_cons(4) do |a, b, c, d|
        if b + c + d > a + b + c
            counter += 1
        end 
    end
    return counter
end

puts "Part 1 - Test: " + partOne(testInputFile).to_s
puts "Part 1 - Actual: " + partOne(puzzleInputFile).to_s
puts "Part 2 - Test: " + partTwo(testInputFile).to_s
puts "Part 2 - Actual: " + partTwo(puzzleInputFile).to_s